### PR TITLE
Return 200 OK response code for a cluster health timeout

### LINKED ISSUE #70849: Fix HTTP response code for cluster health API
Today `GET _cluster/health` returns `408 Request timeout` if it times out before the desired condition is reached. This response code is to indicate that the server timed out waiting for a request from the client, so it isn't appropriate here at all. There's no great fit for a server-side timeout response code, I suggest `503`.

This API already returns `503 Service unavailable` if there is no master, after waiting a while, but immediately returns `200 OK` if there is a master but `STATE_NOT_RECOVERED_BLOCK` is in place. I think we should wait and return `503` in the presence of the `STATE_NOT_RECOVERED_BLOCK` too.

### PR DESCRIPTION
**Pull request by arteam**
Returning 408 for a cluster health timeout was deprecated in #78180 and backported to 7.x in #78940

Now we can do a breaking change in 8.0 respecting the user choice to run ES in 7.x compatible mode via the REST Compatibility layer. 

Fixes #70849

### CODE REVIEW

**Comment by arteam** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine update branch

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I left some suggestions & comments.

**Review by DaveCTurner — COMMENTED** [review]
Thanks @arteam, I left a couple of comments.

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine run REPRODUCE WITH: ./gradlew ':qa:mixed-cluster:v8.0.0#mixedClusterTest' -Dtests.class="org.elasticsearch.backwards.MixedClusterClientYamlTestSuiteIT" -Dtests.method="test {p0=cluster.health/20_request_timeout/cluster health request timeout waiting for active shards}" -Dtests.seed=8E297F5034C5B141 -Dtests.bwc=true -Dtests.locale=ms-MY -Dtests.timezone=America/Kralendijk -Druntime.java=17

**Comment by arteam** [comment]
@elasticmachine run elasticsearch-ci/bwc

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterHealthAction.java [hunk 1]
```diff
@@ -81,9 +88,15 @@ public static ClusterHealthRequest fromRequest(final RestRequest request) {
         if (request.param("wait_for_events") != null) {
             clusterHealthRequest.waitForEvents(Priority.valueOf(request.param("wait_for_events").toUpperCase(Locale.ROOT)));
         }
-        clusterHealthRequest.return200ForClusterHealthTimeout(
-            request.paramAsBoolean("return_200_for_cluster_health_timeout", clusterHealthRequest.doesReturn200ForClusterHealthTimeout())
-        );
+        String return200ForClusterHealthTimeout = request.param("return_200_for_cluster_health_timeout");
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterHealthAction.java:91]: Please use a named constant for the param name both here and in the deprecation message to make it clear it's not just a coincidence that they're the same string. It's ok to use the literal string in tests.

File: rest-api-spec/build.gradle [hunk 2]
```diff
@@ -82,6 +82,7 @@ tasks.named("yamlRestTestV7CompatTransform").configure { task ->
   task.skipTest("search.aggregation/20_terms/string profiler via map", "The profiler results aren't backwards compatible.")
   task.skipTest("search.aggregation/20_terms/numeric profiler", "The profiler results aren't backwards compatible.")
   task.skipTest("migration/10_get_feature_upgrade_status/Get feature upgrade status", "Awaits backport")
+  task.skipTest("cluster.health/20_request_timeout/cluster health request timeout with 200 response code", "Backwards incompatible, returns a warning in 8.0")
```
Comment (DaveCTurner) [inline rest-api-spec/build.gradle:85]: I think we shouldn't be adding to the list of skipped tasks. Instead, we can use the `allowed_warnings` feature to say that this warning is allowed, but not required, in the compatibility tests.

File: docs/changelog/78968.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+pr: 78968
+summary: HTTP Status code has changed for the Cluster Health API in case of a server timeout
```
Comment (DaveCTurner) [inline docs/changelog/78968.yaml:2]: This duplicates the entry in `docs/reference/migration/migrate_8_0/cluster.asciidoc` - are you sure both are needed?

**Review by DaveCTurner — COMMENTED** [review]
Looks good, just a few questions and comments.

File: rest-api-spec/build.gradle (hunk 2 above)
Comment (arteam) [reply rest-api-spec/build.gradle:85]: Created https://github.com/elastic/elasticsearch/pull/80178 to backport the change to 7.16 and then I can enable the test again, so and the BWC tests won't be broken.

File: docs/changelog/78968.yaml (hunk 3 above)
Comment (arteam) [reply docs/changelog/78968.yaml:2]: I'm not sure, but I got a warning from @elasticmachine that I need to create a changelog entry when I add a `breaking` label to the PR.

**Comment by arteam** [comment]
@elasticmachine update

**Comment by arteam** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine run elasticsearch-ci/bwc

File: docs/changelog/78968.yaml (hunk 3 above)
Comment (DaveCTurner) [reply docs/changelog/78968.yaml:2]: Yes I've not seen this before, I suggest checking with delivery or docs folks what to do about this.

**Review by DaveCTurner — APPROVED** [review]
This LGTM except for the question about the duplicated docs.

**Comment by arteam** [comment]
@elasticmachine update branch

**Pull request merged**